dev/list_failed_cross_version_tests.py

```diff
@@ -0,0 +1,61 @@
+"""
+A script to list failed jobs in the latest scheduled run of the `cross-version-tests` workflow
+using the GitHub Actions API.
+
+References:
+- https://docs.github.com/en/rest/reference/actions#list-workflow-runs
+- https://docs.github.com/en/rest/reference/actions#list-jobs-for-a-workflow-run
+"""
+import json
+import os
+import requests
+
+
+def fetch_failed_jobs(session, run_id):
+    failed_jobs = []
+    per_page = 100
+    page = 1
+    while True:
+        url = f"https://api.github.com/repos/mlflow/mlflow/actions/runs/{run_id}/jobs"
+        resp = session.get(url, params={"per_page": per_page, "page": page})
+        jobs = resp.json()["jobs"]
+        failed_jobs.extend(j for j in jobs if j["conclusion"] == "failure")
+        if len(jobs) < per_page:
+            break
+        page += 1
+
+    return failed_jobs
+
+
+class StrictSession(requests.Session):
+    """
+    A wrapper class for `requests.Session` to validate the request is successful.
+    """
+
+    def request(self, *args, **kwargs):
+        resp = super().request(*args, **kwargs)
+        resp.raise_for_status()
+        return resp
+
+
+def create_session():
+    session = StrictSession()
+    token = os.environ.get("GITHUB_TOKEN")
+    if token:
+        session.headers.update({"Authorization": f"token {token}"})
```

Comment:
this throws if someone doesn't have a token with authorization to get job results, right?

Reply:
You can run this script with a token, but you will hit the rate limit pretty quickly:

https://docs.github.com/en/rest/overview/resources-in-the-rest-api#rate-limiting

> When using the built-in GITHUB_TOKEN in GitHub Actions, the rate limit is 1,000 requests per hour per repository. For organizations that belong to a GitHub Enterprise Cloud account, this limit is 15,000 requests per hour per repository.

> For unauthenticated requests, the rate limit allows for up to 60 requests per hour. Unauthenticated requests are associated with the originating IP address, and not the user making requests.